Add Mono EventPipe rundown support.

lateralusX opened the pull request:
Add support into Mono VM emitting rundown events into EventPipe stream. All rundown events emitted by CoreClr during EventPipe rundown phase are now also emitted by Mono, making sure enough meta information is available for tooling to correctly resolve callstacks (not yet emitted) included in EventPipe events emitted by Mono VM.

Re-eanble couple of runtime tests disabled on Mono due to lack of rundown events.

lateralusX committed:
Add Mono EventPipe rundown support.

Add support into Mono VM emitting rundown events into EventPipe stream.
All rundown events emitted by CoreClr during EventPipe rundown phase
are now also emitted by Mono, making sure enough meta information
is available for tooling to correctly resolve callstacks (not yet emitted)
included in EventPipe events emitted by Mono VM.

lateralusX committed:
Enable additional runtime tests passing after emitting rundown events.

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -271,6 +278,1245 @@ ep_rt_mono_os_environment_get_utf16 (ep_rt_env_array_utf16_t *env_array)
+
+static
+bool
+write_runtime_info_dc_start (
```
I believe the equivalent of these functions in coreclr are autogenerated at build time from a manifest.  Rather than manually curating these, could we re-use the scripts that auto-generate them for coreclr?  It would ensure that they are in sync on both ends.  Preferably, we might use the same generated code, but the generated code uses C++ features I believe, so we'd have to modify them.

lateralusX replied:
Yes that is the long term goal, but wanted to get rundown working first on Mono and then spend time on converting the script into something that can be consumed by both runtimes.

josalem replied:
That sounds like a decent plan.

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.h:
```diff
@@ -1308,7 +1330,11 @@ inline
 ep_rt_execute_rundown (void)
 {
-	//TODO: Implement.
+	if (ep_rt_config_value_get_rundown () > 0) {
```
You should be able to also request rundown on a per-session basis.  This logic will either disable or enable globally only using the env var.

lateralusX replied:
This is the same logic as coreclr,. so this is per session at call site, but there is currently an option to disable to globally as well, see https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/vm/eventpipesession.cpp#L610 for coreclr C++ implementation and https://github.com/dotnet/runtime/blob/16e6e9530687b34cc93a85d4043093f06f9b7eff/src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h#L1877 for coreclr C usage.

lateralusX replied:
and here is the logic for this at call site:

https://github.com/dotnet/runtime/blob/e29b83934e7ceef9905f82bd6328cbc72db2278b/src/native/eventpipe/ep.c#L529

so each session can dictate the need for rundown, but the global property can disable it for all (but have no history why that was implemented to begin with).

josalem replied:
The global flag is most likely a holdover from when EventPipe was a single session started by an environment variable or a config file placed at a known location.  We may want to look at removing that eventually.

lateralusX committed:
Implement custom alloc for utf8/utf16 conversion.

Reduce heap allocations for all strings emitted in rundown. Support
buffer in place conversions.

lateralusX committed:
Move assembly/method rundown iteration into runtime source.

lambdageek commented on src/mono/mono/eglib/glib.h:
```diff
@@ -1004,7 +1004,8 @@ typedef enum {
 	G_CONVERT_ERROR_BAD_URI,
-	G_CONVERT_ERROR_NOT_ABSOLUTE_PATH
+	G_CONVERT_ERROR_NOT_ABSOLUTE_PATH,
+	G_CONVERT_ERROR_ALLOC_FAILED
```
gtk glib uses `G_CONVERT_ERROR_NO_MEMORY` - https://developer.gnome.org/glib/stable/glib-Character-Set-Conversion.html#GConvertError maybe we shoudl use the same name.

lambdageek commented on src/mono/mono/eglib/glib.h:
```diff
@@ -1031,6 +1032,32 @@ size_t     g_utf16_len     (const gunichar2 *);
+
+static
+gpointer
+g_converter_default_custom_allocator_func (gsize req_size, gpointer custom_alloc_data)
```
nit: it's not the default, the default is `NULL`.  This is some kind of fixed buffer custom allocator.

lateralusX replied:
Was moving back and forth on this, had fixed_buffer at one point, but then moved back to the default custom allocator, so not the default allocator, but maybe should use a different naming to make things even more clear.

lambdageek commented on src/mono/mono/eglib/glib.h:
```diff
@@ -1031,6 +1032,32 @@ size_t     g_utf16_len     (const gunichar2 *);
+	return default_custom_alloc_data->buffer;
+}
+
+gunichar2 *g_utf8_to_utf16_custom_alloc (const gchar *str, glong len, glong *items_read, glong *items_written, GConvertCustomAllocator custom_alloc_func, gpointer custom_alloc_data, GError **err);
+gchar *g_utf16_to_utf8_custom_alloc (const gunichar2 *str, glong len, glong *items_read, glong *items_written, GConvertCustomAllocator custom_alloc_func, gpointer custom_alloc_data, GError **err);
```
these (and the other new symbols) need to also be added to `eglib-remap.h` so that we don't clash with gtk glib

lambdageek commented on src/mono/mono/metadata/icall-eventpipe.c:
```diff
@@ -132,6 +248,249 @@ eventpipe_thread_detach (void)
+		method_name = method->name;
+		method_signature = mono_signature_full_name (method->signature);
+
+		if (method->klass) {
```
`method->klass` will be set for wrapper methods too (for some wrapper types it's arbitrarily set to `System.Object`). Is that reasonable?

lateralusX replied:
I have seen some wrappers in CoreClr's events as well, so I believe we could start with that and adjust later if we hit issues.

lambdageek commented on src/mono/mono/metadata/domain-internals.h:
```diff
@@ -665,6 +665,11 @@ void mono_assembly_cleanup_domain_bindings (guint32 domain_id);
+typedef void (*MonoJitInfoFunc) (MonoJitInfo *ji, gpointer user_data);
+
+void
+jit_info_table_foreach (MonoDomain *domain, MonoJitInfoFunc func, gpointer user_data);
```
nit: `mono_jit_info_table_foreach`

lateralusX replied:
OK, will probably add _internal to it as its used for naming of similar methods.

lambdageek reviewed (commented):
Mono bits mostly look ok. A couple of nits. Really the only important bit is to add new glib.h symbols to eglib-remap.h, too.

lateralusX committed:
Review feedback.

lateralusX committed:
Remove duplicated test exclusions.

lateralusX committed:
Fix build warnings + additional name adjustments.

lateralusX committed:
Use eg_ instead of g_ prefix for new static function.

josalem commented:
Have you tried collecting a trace using `dotnet-trace` then opening it in PerfView or VS?  I imagine that the end-to-end success criteria for this should be getting stacks for a simple trace.

lateralusX commented:
> Have you tried collecting a trace using `dotnet-trace` then opening it in PerfView or VS? I imagine that the end-to-end success criteria for this should be getting stacks for a simple trace.

Collecting stack traces is not yet implemented on Mono. Plan was to build on top of this and implement sample profiler + stack traces and when that is done I will run end-to-end test. I have matched the event and data emitted by CoreClr during rundown in this PR, so most of the information should already be included in the stream once the stack traces are added into the mix. If there are issues detected parsing stack traces due to rundown events I will fix that in upcoming PR's.
